python/ray/experimental/data/block.py

```diff
@@ -6,9 +6,26 @@
     import pyarrow
     from ray.experimental.data.impl.block_builder import BlockBuilder
 
+from ray.util.annotations import DeveloperAPI
+
 T = TypeVar("T")
 
 
+@DeveloperAPI
+class Block(Generic[T]):
```

Comment:
Open to suggestions on a better way to do this... if we went with Union[pyarrow.Table, list], that would work, but we'd lose the Generic[T].

Reply:
I'd vote for `Block = Union[pyarrow.Table, List[Any]]` being the lesser of two evils, since it actually gives the real types of the block while this generic dummy type doesn't have any real relation to the actual content types, and the type variable `T` wouldn't apply to the `pyarrow.Table` case anyway. E.g., although `BlockAccessor.iter_rows()` returns `Iterator[T]` on the base class, `ArrowBlockAccessor.iter_rows()` returns `Iterator[ArrowRow]` since `ArrowBlock` doesn't follow the `Block[T]` type scheme.

The output type for `BlockAccessor.iter_rows()` will probably be `Iterator[Any]`, but that will be more accurate than `Iterator[T]` anyway.

Reply:
Let's keep thinking about this then. Block[ArrowRow] is semantically correct actually, since you iterate over ArrowRow objects there.

I'm inclined to leave this PR as is since it's already implementing one reasonable option.

Reply:
@ericl Sure I'm fine with going ahead with this, we can move to a better option later.

Reply:
Changed to Union[List[T], _ArrowTable[T]]